### ISSUE TITLE
Consistent way to publish pytorch and custom models using zenodo for data storage

### ISSUE THREAD

**Issue by krrome** [kipoi/kipoi#339]
Multiple files may be required by models, which are currently not mentioned in the `model.yaml`. an option would be to add a field to the `model.yaml` and adding a `destination` field to the URL defintion:

```yaml
type: pytorch
args:
  file:
    url: http://.../model_files/deepsea_cpu.py
    md5: 68vf768s...
  build_fn: get_seqpred_model
extra_files:
  - url: http://...
    md5: 576sdtv7...
    destination: model_files/filename # subfolder and filename how it should be stored
  - url: http://...
    md5: 9gdfb578...
    destination: model_files/filename2 # subfolder and filename how it should be stored
```

**Comment by Avsecz** [comment]
@krrome why don't we just move deepsea_cpy to git? It's a python file anyway?

**Comment by krrome** [comment]
regarding pytorch: we don't necessarily have an argument in which we could specify the URL. as Pytorch is so general that it doesn't have and proper "reload_a_model()" function in their API. Hence there is no "standard way" of doing that.

But in general, when people want to use their custom models it will be necessary for them download an arbitrary number of files into the correct location. We could shift that task to the contributor, requiring them to download the files themselves in their Model class, but that I think is not very nice + you do all the md5sum checking in a very nice and central way, so we should extend that for all the files required for any model to run. or am I missing something there.

**Comment by Avsecz** [comment]
Is there a way for the user to write a function that just compiles the model and then we can populate it by downloading weights?

**Comment by Avsecz** [comment]
From https://pytorch.org/tutorials/beginner/saving_loading_models.html:

![image](https://user-images.githubusercontent.com/10011161/46730434-6d874400-cc88-11e8-83c4-92d8a50cef8b.png)

For example, the user could provide the model class and we could run `load_state_dict` on it

**Comment by krrome** [comment]
ok to be fair the last time I read the pytorch docs was when I implemented the PyTorch Model class, and back then they explicitly said one should not use the `load_state_dict`. I guess a lot has happened in the meantime. I guess this means we can limit ourselves now to that.

but what about all the custom models? what about additional files that are needed for dataloaders? That is already the case for your contributing_models.ipynb

**Comment by Avsecz** [comment]
Let's switch to the new pytorch way. It will be much easier and they released 1.0 anyway.

There are two ways how to deal with custom classes:

#341 ## Custom models: Solution 1
models accept args a path to the weights file and then we specify them in the yaml file 

```python
class MyModel:
  def __init__(self, weights_file):
    pass
```
```yaml
type: custom
args:
  path: model.py...
  args:
    weights_file:
      url: ....
      md5: 
```

### Custom models: Solution 2

```python
class MyModel:
  def __init__(self):
    from kipoi.external.keras.data_utils import get_file
    weights_file = get_file("...", )
    ...
```

**Comment by krrome** [comment]
Solution 1 is trivial, but solution 2 could potentially work for dataloaders. or is there a different way you have in mind to assign a file path to the dataloader that is always the same? e.g. have an init argument with a file path, where the default is the file path? It can't be an `example`...

**Comment by krrome** [comment]
what I don't like about solution 2 is that:
1) the kipoi API may change
2) the model.yaml does not list all the external files

**Comment by Avsecz** [comment]
> the model.yaml does not list all the external files

I fully agree about this. Let me think about it

**Comment by Avsecz** [comment]
How about the following API change:

Dataloder 
```python
type: Dataset  # optional if it's a python generator or a function
defined_as: dataloader.SplicingKmerDataset  # Changed from dataloader.py::SplicingKmerDataset
args:
    gtf_file:
        doc: file path; Genome annotation GTF file
        example: example_files/hg19.chr22.gtf
    fasta_file:
        doc: Reference Genome sequence in fasta format
        example: example_files/hg19.chr22.fa  # used for testing
        default:  # used as a default on normal runtime   #  added attribute
           url: path_to_fasta
           md5: ....
```

Custom model
```yaml
defined_as: model.HALModel
args:
  weights:
    url: url
    md5:
  other_arg:
    doc: description
    default:
      url: ...
      md5: ....
```

Keras model:
```yaml
defined_as: kipoi.model.KerasModel
args:
  weights:
     url: ...
```

**Comment by krrome** [comment]
I like that, this way we require contributors to be explicit about their external dependencies.

**Comment by Avsecz** [comment]
closed by #348 and #346
